### PR TITLE
Do not require FC header for /status [GAWB-2865]

### PR DESCRIPTION
**Pull request by aednichols**
The `/status` endpoint shouldn't require the `X-FireCloud-Id` header. Right now it does.

Without the headers provided by Swagger (e.g. a load balancer's healthcheck function), requests to the status endpoint seem to get routed into `fireCloudProtectedRoutes` and then rejected. If we just move status up in the eval order, things work as expected.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by rushtong** [comment]
This requires PO/FISMA/security approval. The FC ID header was added specifically to prevent access from any non-FC system.

**Review by dvoet — COMMENTED** [review]
The order should not matter and if it does then that is the actual bug.

**Comment by davidangb** [comment]
I don't believe Thurloe's /status requires the `X-FireCloud-Id` header. I can hit it with swagger or curl and get a response just fine.

**Comment by aednichols** [comment]
@davidangb, Bernick and Rob M. ran into an issue with a load balancer healthcheck that doesn't let you pass parameters:
```
curl https://thurloe201.dsde-dev.broadinstitute.org/status
```
returns
```
Missing 'X-FireCloud-Id' Header
```

@rushtong I think the status endpoint is protected only by accident; it's not in `fireCloudProtectedRoutes`, and the tests for it do not have a `addHeader` step. Does that change the required formalities? (I can see why it might not.)

File: src/main/scala/thurloe/service/ThurloeServiceActor.scala
Comment (rushtong) [inline src/main/scala/thurloe/service/ThurloeServiceActor.scala]: @aednichols - Can you test this without the swap? If it still works in the original, please revert.

Comment (rushtong) [reply src/main/scala/thurloe/service/ThurloeServiceActor.scala]: Ignore this, sorry.

File: src/main/scala/thurloe/service/ThurloeServiceActor.scala
Comment (rushtong) [inline src/main/scala/thurloe/service/ThurloeServiceActor.scala]: Revert please :-)

**Review by rushtong — APPROVED** [review]
👍

**Comment by aednichols** [comment]
Talked to @davidbernick and he's cool with this for FISMA purposes (tagged him as reviewer as a courtesy).

**Comment by aednichols** [comment]
I fixed the ignored test that let us miss this (by means of mock DAOs).

**Review by davidangb — CHANGES_REQUESTED** [review]
do not merge. There's something else fishy going on beyond the X-FireCloud-Id header, and we need to get to the bottom of it to understand the proper change to make.
```
curl https://thurloe201.dsde-dev.broadinstitute.org/status

Missing 'X-FireCloud-Id' Header
```
but:
```
curl -X GET --header "Accept: application/json" "https://thurloe201.dsde-dev.broadinstitute.org/status"

{"status": "up"}
```

possibly we are returning the "Missing 'X-FireCloud-Id' Header" error message in a case we shouldn't be?

Try this through swagger: https://thurloe.dsde-dev.broadinstitute.org/swagger/index.html?url=/swagger/thurloe.yaml#!/Status/status ... the status endpoint works just fine without the header. Note that swagger *does* send the `Accept` header. I realize the health check can't send the header, but we still need to figure out the right fix.

**Comment by davidangb** [comment]
After running this locally, I see that your changes do in fact fix the issue. Could you explain them?

**Comment by aednichols** [comment]
Sure. When we had
```
val fireCloudProtectedRoutes: Route = optionalHeaderValueByName(fcHeader) { [...] }
```
sending a `/status` request with no headers matched this route because the header is optional, and because this route comes before `statusRoute` in `ThurloeServiceActor.scala`.

Once matched, we fell into the `None` case. You can verify this by changing (off of `develop`) line `18` of `FireCloudProtectedServices.scala` to
```
case None => respondWithStatus(StatusCodes.BadRequest) { complete("Jimmy") }
```
```
$ curl https://local.broadinstitute.org:40443/status
Jimmy
```

It doesn't really make sense to allow header-less requests to "fall into" this route only to immediately fail as a bad request, so I switched to `headerValueByName`. From its documentation,
```
If no header with a matching name is found the request is rejected with a [[spray.routing.MissingHeaderRejection]].
```

This causes the route not to match if there's no header, and Spray evaluation proceeds down to `statusRoute`.

**Comment by aednichols** [comment]
The one thing I'm not sure about is what will happen when we send a `fireCloudProtectedRoutes` request with no FC header - will we get a useful error message? Will test with curl.

**Review by davidangb — APPROVED** [review]
:+1:

**Review by davidangb — APPROVED** [review]
:+1: once Travis tests pass

**Pull request merged**
